controllers/bsl.go

```diff
@@ -416,16 +422,23 @@ func (r *DPAReconciler) ensureSecretDataExists(dpa *oadpv1alpha1.DataProtectionA
 				return fmt.Errorf("Secret name specified in BackupLocation %s cannot be empty", bsl.Name)
 			}
 		}
+
 		// Check if the BSL secret key configured in the DPA exists with a secret data
-		secretName, secretKey := r.getSecretNameAndKeyforBackupLocation(*bsl)
-		bslSecret, err := r.getProviderSecret(secretName)
-		if err != nil {
-			return err
+
+		if bsl.CloudStorage != nil {
+			_, _, err := r.getSecretNameAndKeyFromCloudStorage(bsl.CloudStorage)
+			if err != nil {
+				return err
+			}
 		}
-		data, foundKey := bslSecret.Data[secretKey]
-		if !foundKey || len(data) == 0 {
-			return fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
+
+		if bsl.Velero != nil {
+			_, _, err := r.getSecretNameAndKey(bsl.Velero.Config, bsl.Velero.Credential, oadpv1alpha1.DefaultPlugin(bsl.Velero.Provider))
```

Comment:
I think this function should be removed from here (and vsl). The idea is to check for errors, right? It can be done direct in velero `if` with this, not? (which avoid checking again if name and key are not empty)
```go
		if bsl.Velero != nil && bsl.Velero.Credential != nil {
			...
			err := r.verifySecretContent(bsl.Velero.Credential.Name, bsl.Velero.Credential.Key)
			if err != nil {
				return err
			}
		}
```

Reply:
same logic will be written 2 times then..

Reply:
this getSecretNameandKey from velero credentials, is also being used at multiple places, so i have made it as a common function

Reply:
Also @mpryc suggested that, the function getting the secret name and key should also validate the content of the secret, hence the return type of the function is changed to (string, string, error)

Reply:
But like this, we would be checking the same thing multiple times, not?

here would be the necessary check
https://github.com/openshift/oadp-operator/blob/91e1aacf47ea2b44bafdb9d6d3c14ffee6907baf/controllers/dpa_controller.go#L95

and then here, instead of just using the validated data, I would validate it again...
https://github.com/openshift/oadp-operator/blob/91e1aacf47ea2b44bafdb9d6d3c14ffee6907baf/controllers/dpa_controller.go#L97

Reply:
@mateusoliveira43 thanks for the clarification! I understand the ensureSecretDataExists() function is being called in both ValidateBackupStorageLocations & ReconcileBackupStorageLocations.

So, my question is , can we have a scenario where only one function executes out of the 2, like on updating BSL directly? Or no?

@sseago @kaovilai @shubham-pampattiwar anyone can please clarify?

Reply:
ok, lets discuss if it is fine to check the content twice.

Reply:
Twice is fine as long as it's not too much duplicate code otherwise make it a function.

Reply:
not much duplicate i believe

Reply:
There are 3 options:
- use a `skipVerify` flag in the function
- do not use a function
- make peace with extra verification

Reply:
I am OK with either 1 or 3

Reply:
what do you suggest @shubham-pampattiwar

Reply:
I am good with either of the approaches I listed. This is not a big concern for me. I am always in favor of simple, readable and maintainable code.